protocols/kad/src/behaviour.rs

```diff
@@ -922,21 +938,47 @@ where
     /// The result of this operation is delivered in a
     /// reported via [`KademliaEvent::OutboundQueryCompleted{QueryResult::GetProviders}`].
     pub fn get_providers(&mut self, key: record::Key) -> QueryId {
-        let providers = self
+        let providers: HashSet<_> = self
             .store
             .providers(&key)
             .into_iter()
             .filter(|p| !p.is_expired(Instant::now()))
             .map(|p| p.provider)
             .collect();
+
         let info = QueryInfo::GetProviders {
             key: key.clone(),
-            providers,
+            providers_found: providers.len(),
+            step: ProgressStep::first(),
         };
-        let target = kbucket::Key::new(key);
+
+        let target = kbucket::Key::new(key.clone());
         let peers = self.kbuckets.closest_keys(&target);
         let inner = QueryInner::new(info);
-        self.queries.add_iter_closest(target.clone(), peers, inner)
+        let id = self.queries.add_iter_closest(target.clone(), peers, inner);
+
+        // No queries were actually done for the results yet.
+        let stats = QueryStats::empty();
+
+        if !providers.is_empty() {
+            self.queued_events
+                .push_back(NetworkBehaviourAction::GenerateEvent(
+                    KademliaEvent::OutboundQueryProgressed {
+                        id,
+                        result: QueryResult::GetProviders(Ok(GetProvidersOk::FoundProviders {
+                            key,
+                            providers,
+                        })),
+                        step: ProgressStep::first(),
+                        stats,
+                    },
+                ));
+            let query = self.query_mut(&id).expect("just inserted");
+            if let QueryInfo::GetProviders { step, .. } = &mut query.query.inner.info {
+                *step = step.next();
+            }
```

Comment:
Same here. Altering this after creation is confusing to me. I suggest to do the same change as done for `get_record`. See 5e8f3cce8.